### PR TITLE
[tune] refactor search spaces (old)

### PR DESCRIPTION
**Pull request by krfricke**
# After a failed rebase the discussion has been moved here: #10444

## Why are these changes needed?

This introduces a new search space representation that makes it possible to convert a Tune search space to other search algorithm definitions.

This also introduces new sampling methods, like quantized variants `uniform` and `loguniform`, called `quniform` and `qloguniform`, respectively.

With these abstractions we get a natural way to distinguish between allowed parameter values (called `Domain`s) and the sampling methods (e.g. uniform, loguniform, normal). Theoretically users can introduce their own domains and custom samplers (like sampling from a Beta distribution or so). The underlying API is quite flexible, e.g. `Float(1e-4, 1e-2).loguniform().quantized(5e-3)`. This API is currently hidden behind the tune sampler functions, like `tune.qloguniform(1e-4, 1e-2, 5e-3)`.

Converting Tune search space definitions to search spaces for external search algorithms, like AxSearch, HyperOpt, BayesOpt, etc. ist straightforward. If a search algorithm doesn't support specific sampling methods, they can be dropped with a warning, or an error can be raised. For instance, BayesOpt doesn't support custom sampling methods, and is only interested in parameter bounds. If someone passes `Float(1e-4, 1e-2).qloguniform(5e-3)` to BayesOpt, it will be converted to the parameter bounds `(1e-4, 1e-2)` and a warning will be raised stating that the custom sampler has been dropped.

Generally, this refactoring will introduce flexibility in defining and converting search spaces, while keeping full backwards compatibility.

# Example usage:

External API:

```python
config = {
    "a": tune.choice([2, 3, 4]),
    "b": {
        "x": tune.qrandint(0, 5, 2),
        "y": 4,
        "z": tune.loguniform(1e-4, 1e-2)
    }
}
converted_config = HyperOptSearch.convert_search_space(config)
```

Lower-level API equivalent:

```python
config = {
    "a": tune.sample.Categorical([2, 3, 4]).uniform(),
    "b": {
        "x": tune.sample.Integer(0, 5).quantized(2),
        "y": 4,
        "z": tune.sample.Float(1e-4, 1e-2).loguniform()
    }
}
converted_config = HyperOptSearch.convert_search_space(config)
```

## Related issue number

Concerns #9969

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

File: python/ray/tune/suggest/optuna.py [hunk 1]
```diff
@@ -147,3 +161,64 @@ def restore(self, checkpoint_path):
...
+                "Optuna search does not support parameters of type "
+                "`{}` with samplers of type `{}`".format(
+                    type(domain).__name__,
+                    type(domain.sampler).__name__))
+
+        values = []
+        for path, domain in domain_vars:
+            par = "/".join(path)
+            values.append(resolve_value(par, domain))
+        return values
```
Comment (sumanthratna) [inline python/ray/tune/suggest/optuna.py:224]: you know what I'm going to say :)

File: python/ray/tune/utils/util.py [hunk 2]
```diff
@@ -231,6 +231,18 @@ def flatten_dict(dt, delimiter="/"):
...
 
 
+def unflatten_dict(dt, delimiter="/"):
+    """Unflatten dict. Does not support unflattening lists."""
+    out = defaultdict(dict)
+    for key, val in dt.items():
+        path = key.split(delimiter)
+        item = out
+        for k in path[:-1]:
+            item = item[k]
```
Comment (sumanthratna) [inline python/ray/tune/utils/util.py:241]: can't this be shortened to `item = item[path[-2]]`? I think I'm misunderstanding something here

**Review by sumanthratna — COMMENTED** [review]
this is amazing! I left a bunch of comments about list/dict comprehensions which might improve perf but aren't really necessary and might decrease code readability

also, I wonder if we should create a `tune.set_seed` function to set the seed for numpy and python, similar to [`pytorch_lightning.seed_everything`](https://pytorch-lightning.readthedocs.io/en/latest/trainer.html#reproducibility)

File: python/ray/tune/suggest/optuna.py [hunk 3]
```diff
@@ -91,6 +102,8 @@ def __init__(
 
         self._space = space
 
+        self._config = config or {}
```
[the remark this answers is not in the document]
Comment (krfricke) [reply python/ray/tune/suggest/optuna.py:105]: Throws a PEP8 error for me (default argument is mutable)

File: python/ray/tune/utils/util.py (hunk 2 above)
Comment (krfricke) [reply python/ray/tune/utils/util.py:241]: Path here is usually a tuple like `(a, b, c)`. We would like to set `dt[a][b][c]=x` and do this by setting `item=dt`, then `item = item[a]`, then `item=item[b]` and finally `item[c]=x`. Does that make sense?
See also `assign_value()` in `variant_generator.py`

**Comment by krfricke** [comment]
Thanks for the benchmarks and suggestions, I applied them.

**Comment by krfricke** [comment]
Oh well, something went wrong with the latest merge. I'll try to fix that

**Comment by krfricke** [comment]
Weird, when I create a PR from the exact same branch it doesn't blow. Let's continue the discussion here: #10444

**Pull request closed without merging**